### ISSUE TITLE
experiment with converting quadratic functions to nonlinear, to simplify/streamline code

### ISSUE THREAD

**Issue by chriscoey** [jump-dev/Pavito.jl#67]
Pavito is harder to maintain, debug, and read/understand while the quadratic functions are handled separately from the more general nonlinear functions. 

after ScalarNonlinearFunction is finalized and https://github.com/jump-dev/MathOptInterface.jl/pull/2059 is in, I think we should think about perhaps a bridge from ScalarQuadraticFunction to ScalarNonlinearFunction in MOI, or if not a bridge then a more manual conversion inside Pavito. 

@odow any thoughts? I'd like to have a shot at this when the time is right.

**Comment by chriscoey** [comment]
Hmm I wonder if https://github.com/jump-dev/Ipopt.jl/pull/320 slash https://github.com/jump-dev/MathOptInterface.jl/issues/1397 may help here.

**Comment by blegat** [comment]
Yes, I agree that having a bridge from ScalarQuadraticFunction to ScalarNonlinearFunction would be ideal and would simplify Pavito but it seems to be hard to handle constraints one by one, see https://github.com/jump-dev/MathOptInterface.jl/pull/2059#issuecomment-1355618965

**Comment by chriscoey** [comment]
I'm a little confused about that point. if our problem has a bunch of ScalarQuadraticFunctions constraints and ScalarNonlinearFunction constraints, and we bridge the SQFs to SNFs, then we have a problem with all SNFs. and that pure nonlinear model can be handled with a single NLP block - so all grad, Hess etc oracles can be done in one go

**Comment by chriscoey** [comment]
Ok, I'd be surprised if it made much difference in practice, but yeah @odow would have a better idea than me of course.

**Comment by ccoffrin** [comment]
One thing to keep an eye out for is derivative performance on very large
problems.  Jacobian and Hessian computations can be much faster for the
subset of linear and quadratic expressions than general nonlinear
expressions.

**Comment by chriscoey** [comment]
This would be an issue for the nonlinear solver, so we would want to preserve quadraticness there. but for deriving cuts from gradients for the OA mip solver, it shouldn't matter because the costs would surely be dwarfed.

**Comment by odow** [comment]
> I think we should think about perhaps a bridge from ScalarQuadraticFunction to ScalarNonlinearFunction in MOI

Yes. That's reasonable once we have an ScalarNonlinearExpression in MOI.

> Jacobian and Hessian computations can be much faster for the subset of linear and quadratic expressions than general nonlinear expressions.

The default JuMP AD has support for detecting linear expressions, but not quadratic.

> but for deriving cuts from gradients for the OA mip solver, it shouldn't matter because the costs would surely be dwarfed.

Yeah. For the trade-off, the overhead of Pavito going back and forward to the MIP solver, etc, is probably larger than the AD cost of explicit quadratic versus an NLP version of quadratic.

**Comment by odow** [comment]
8 months later... we should take another look at this

**Comment by chriscoey** [comment]
Yeah! I won't have time in the next few months I think, but happy if anyone else wants to try it
